### PR TITLE
Significantly reduce SafeX509Handle finalizations

### LINKED ISSUE #9143: Dispose unneeded certs in chain walk
Followup to #8978.

OpenSslChainProcessor.FindCandidates causes a lot of X509Certificate2 objects to be spawned (https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs#L495-L498).  When the toProcess queue is empty it should find all certificates not referenced in `candidates` and Dispose() them, to be a much better citizen with the underlying SafeX509Handles.

### PR DESCRIPTION
**Pull request by bartonjs**
I hooked some basic instrumentation up to the SafeHandle types used in OpenSSL interop, and ran the tests.  The glorious 5 second library became a 300 second library, and reported over 5000 finalizations.  I then systematically drove that down to 0.

* With instrumentation, no other changes: 5501 finalizes
* Clean all the finalizes that are due to the tests not disposing their resources: 5197 (-304)
* Make TryReadX509Der call Dispose on the invalid handle before returning false: 4995 (-202)
* Make PFX reading call Dispose on invalid handles before returning false: 4912 (-83)
* Make calling X509Certificate2Collection with validOnly=true free its X509Chain artifacts: 4909 (-3, it's only tested one time because it's a bit silly)
* Make building X509Chain be more environmentally friendly: 375 (-4534)
* Restructure the LM\Root and LM\CA X509Store handle passing (and Export) to make "never read" (which came from "Open(Write) threw an exception") not send all the certs off to the finalizer: 0 (-375)

There's a pending part 2, which is make sure that we are letting things that should be cleaned get cleaned (aka not leaking memory), but this should definitely help with finalization perf.

Fixes #9143 
cc: @stephentoub @AtsushiKan

### CODE REVIEW

**Comment by bartonjs** [comment]
https://github.com/bartonjs/corefx/commits/openssl_leak_detect has the instrumentation I did.  I made Console.ReadLine() force finalization, then invoked the tests with the `-wait` parameter, since it calls ReadLine :)

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ChainPal.cs [hunk 1]
```diff
@@ -44,35 +45,70 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
...
+                }
+#endif
+
+                if (chain.ChainStatus.Length == 0 && downloaded.Count > 0)
+                {
+                    SaveIntermediateCertificates(chain.ChainElements, downloaded);
+                }
+
+                // Everything we put into the chain has been cloned, dispose all the originals.
+                systemTrusted.DisposeAndClear();
```
Comment (mellinoe) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ChainPal.cs]: Perhaps we could call this `DisposeAllAndClear` to make it clearer that multiple things are disposed. Also, why do we need to clear these temporary sets?

File: src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    internal sealed class ReferenceEqualityComparer<T> : IEqualityComparer<T>
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: Nit: ```where T : class``` ?

File: src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        internal static readonly ReferenceEqualityComparer<T> Instance = new ReferenceEqualityComparer<T>();
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: Nit: we actually already have one of these in the repo:
https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Dynamic/Utils/ReferenceEqualityComparer.cs
Could we consolidate?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs [hunk 4]
```diff
@@ -4,167 +4,74 @@
...
-                case X509ContentType.SerializedCert:
-                case X509ContentType.SerializedStore:
-                    throw new PlatformNotSupportedException(SR.Cryptography_Unix_X509_SerializedExport);
-                default:
-                    throw new CryptographicException(SR.Cryptography_X509_InvalidContentType);
-            }
+            // Dispose is explicitly doing nothing here.
+            // The LM\Root and LM\CA stores are reused after being Disposed, because there's no
+            // point in re-allocating the array to instantiate them every time.
+            // But the interface requires Dispose.
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs]: My initial inclination when I was looking at this previously was to ask whether all of the _certs could be disposed of here.  I believe the answer was "no, they may have been handed out and be in use elsewhere... might be worth noting that here.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs [hunk 5]
```diff
@@ -4,167 +4,74 @@
...
-using Microsoft.Win32.SafeHandles;
+using System.Threading;
 
 namespace Internal.Cryptography.Pal
 {
     internal class CollectionBackedStoreProvider : IStorePal
     {
-        private readonly X509Certificate2[] _certs;
-        private static readonly SafeEvpPKeyHandle InvalidPKeyHandle = new SafeEvpPKeyHandle(IntPtr.Zero, false);
+        private readonly bool _cloneOnRead;
```
Comment (AtsushiKan) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs]: Having a PAL method like IStorePal.CopyTo() clone or not clone based on some boolean passed at construction time feels like a design smell. IStorePal.CopyTo() semantics should be consistent across the board. Can't whoever needs the copying do the copying himself? Why does the Unix Pal need to do this and not the Windows Pal?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportStoreProvider.cs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        }
+
+        public void Dispose()
+        {
+            // Don't dispose any of the resources, they're still owned by the caller.
+            _singleCertPal = null;
+            _certs = null;
+        }
+
+        private static void AssertExportOnly()
```
Comment (AtsushiKan) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportStoreProvider.cs]: I would use the verb "Throw" rather than "Assert" in the name as (1) you're not actually checking a condition here and (2) there's no other indication on the caller side that this method never returns.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportStoreProvider.cs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+                }
+
+                return Interop.Crypto.OpenSslEncode(
+                    handle => Interop.Crypto.GetPkcs7DerSize(handle),
+                    (handle, buf) => Interop.Crypto.EncodePkcs7(handle, buf),
+                    pkcs7);
+            }
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportStoreProvider.cs]: It's probably fine, but I'm confused here by who owns what.  We're creating a SafePkcs7Handle, handing it off to something else, and then disposing of it.  Is that something else duplicating it?  And when cert.Handle is added to the pkcs7, is it being duplicated?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/StorePal.cs [hunk 8]
```diff
@@ -283,7 +273,11 @@ private static void LoadMachineStores()
                 }
             }
 
-            s_machineRootStore = rootStore;
+            var rootStorePal = new CollectionBackedStoreProvider(rootStore, cloneOnRead: true);
+            s_machineIntermediateStore = new CollectionBackedStoreProvider(intermedStore, cloneOnRead: true);
+
+            // s_machineRootStore's nullarity is the loaded-state sentinel, so write it with Volatile.
+            Volatile.Write(ref s_machineRootStore, rootStorePal);
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/StorePal.cs]: Is this volatile write because we're in a double-checked locking situation?  I'm assuming so, as it looks like we're inside of a lock here.  If that's not the case, and we're in an optimistic concurrency situation where multiple threads may be racing to publish their created store, this should instead be an Interlocked.CompareExchange, with the loser disposing of their created state.

**Comment by stephentoub** [comment]
A few comments, but LGTM. Nice job addressing all of these.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ChainPal.cs (hunk 1 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ChainPal.cs]: > Perhaps we could call this DisposeAllAndClear

Sounds good, I'll do that.

> Also, why do we need to clear these temporary sets?

We don't really need to; but I felt better about removing the references.  Having a collection of disposed things always feels weird to me.  I can remove the call to Clear (and that part of the name).

File: src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs (hunk 2 above)
Comment (bartonjs) [inline src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: Ha.  I straight-up copied this from System.Core.  I'll make the change here :)

File: src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs (hunk 3 above)
Comment (bartonjs) [inline src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: > Nit: we actually already have one of these in the repo:

Hm, I must have only looked in coreclr.  I'm happy to share, which is why I put this in Common.  I'll consolidate them.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs (hunk 5 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs]: Yeah, I was considering splitting the IStorePal into Loader/Exporter/SystemStore.

Windows PAL always copies, because it asks CAPI for a PCERT_CONTEXT then wraps that with a new managed object.  Everything in the Unix PAL gets complicated because OpenSSL (P)X509 objects don't have private key associations, so we manage that all in managed code.  This particularly applies to PFX loading; which is where the cloneOnRead is false.

So really it's:
* Export source: Given a valid PAL object, needs to call an underlying API.  Do not Dispose, do not clone.
* Import source: Emit a certificate by whatever means necessary. In Unix emulation of CAPI PCERT_CONTEXT this means passing hydrated X509Certificate2 objects.  Caller will only read once.  Do not Dispose (arguably "dispose unless the caller asked for it already"). do not clone.
* System Store: To match what Windows does, every call to the Certificates property (which uses CopyTo in the PAL) has to return a new X509Certificate2 object.  Clone is required to match existing lifetime semantics.  (Dispose is skipped because we don't support writing to the LM stores; User Stores read hot from the filesystem every time, different interface implementation).

If you agree that there are three different things here I can split the interface to make each IThing have more apparent semantics (my hesitation was "well... I don't really need to bother the Windows half with this...")

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportStoreProvider.cs (hunk 6 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportStoreProvider.cs]: > I would use the verb "Throw" rather than "Assert" in the name

Fair point, will do.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportStoreProvider.cs (hunk 7 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportStoreProvider.cs]: > It's probably fine, but I'm confused here by who owns what.

I'll add some comments.  I'm definitely going to have to investigate Pkcs7AddCertificate to find out why that's safe (it probably internally up-refs).  The reason we can dispose the PKCS7 handle is that we've already called Encode to turn it into the byte[]; so we don't need the native container object anymore.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs (hunk 5 above)
Comment (AtsushiKan) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs]: Yes, splitting the interfaces is the best approach. I just figured that suggestion would be a non-starter due to adding a huge amount of work to an already big PR.

But if you’re willing (either here or in a separate PR), I’ll be delighted with that approach ☺

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/StorePal.cs (hunk 8 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/StorePal.cs]: Yeah, the very beginning of the method asserted that the monitor was held.  I can duplicate the assert to down here to make it more clear.

```C#
if (s_machineRootStore == null)
    lock (s_lock)
        if (s_machineRootStore == null)
            LoadMachineStores();  // <-- you are here
```

**Comment by bartonjs** [comment]
I still need to reconcile the two ReferenceEqualityComparer files; but got the rest of the feedback addressed.  I'll want to squash in "Address feedback" and the pending reconciliation; but left things as new commits for now for ease of followup

**Comment by bartonjs** [comment]
* Pre-squash feedback: https://github.com/dotnet/corefx/pull/9510/commits/5f8b768871f0ea550c61688439b2051ab81fc16c
* Pre-squash IStorePal PAL separation: https://github.com/dotnet/corefx/pull/9510/commits/e9593f56aa6253e39466d5652b93b32602affad4

All feedback has been addressed now (hopefully to satisfaction), and squashed into the relevant commits.  The PAL separation commit description is now way better than the previous "cleaned some finalizations in X509Store" description was.

Double checked, and still at 0 finalizes :smile:

**Comment by stephentoub** [comment]
Test Innerloop CentOS7.1 Release Build and Test please
Test Innerloop OSX Release Build and Test please

File: src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs [hunk 9]
```diff
@@ -2,19 +2,27 @@
...
+        public bool Equals(T x, T y)
+        {
+            return ReferenceEquals(x, y);
+        }
 
-        public int GetHashCode(T obj) => RuntimeHelpers.GetHashCode(obj);
+        public int GetHashCode(T obj)
+        {
+            return RuntimeHelpers.GetHashCode(obj);
+        }
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: Not a big deal, but why did you change these from being expression-bodied members?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+
+namespace Internal.Cryptography.Pal
+{
+    internal class CertCollectionLoader : ILoaderPal
+    {
+        private List<ICertificatePal> _certs;
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs]: readonly?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+
+namespace Internal.Cryptography.Pal
+{
+    internal class CertCollectionLoader : ILoaderPal
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs]: sealed?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            _certs = certs;
+        }
+
+        public void Dispose()
+        {
+            // If there're still certificates, dispose them.
+            _certs?.DisposeAll();
+        }
+
+        public void LoadTo(X509Certificate2Collection collection)
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs]: Maybe "MoveTo" since this seems to have move semantics rather than copy semantics?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs [hunk 13]
```diff
@@ -4,169 +4,40 @@
...
 using System.Collections.Generic;
-using System.Collections.ObjectModel;
 using System.Diagnostics;
-using System.Security.Cryptography;
 using System.Security.Cryptography.X509Certificates;
-using Microsoft.Win32.SafeHandles;
 
 namespace Internal.Cryptography.Pal
 {
     internal class CollectionBackedStoreProvider : IStorePal
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs]: sealed?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportProvider.cs [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal class ExportProvider : IExportPal
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ExportProvider.cs]: sealed?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs [hunk 15]
```diff
@@ -539,6 +539,28 @@ private static X509ChainStatusFlags MapVerifyErrorToChainStatus(Interop.Crypto.X
...
+                var candidatesByReference = new HashSet<X509Certificate2>(
+                    candidates,
+                    ReferenceEqualityComparer<X509Certificate2>.Instance);
+
+                // Dispose any certificates we cloned in, but didn't end up needing.
+                // Since extraStore was provided by users, don't dispose anything it contains.
+                Debug.Assert(storesToCheck.Length > 0, "storesToCheck.Length > 0");
+                Debug.Assert(storesToCheck[0] == extraStore, "storesToCheck[0] == extraStore");
+
+                for (int i = 1; i < storesToCheck.Length; i++)
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: We're skipping the first one because all of its certs were used?  Is it possible there were dups within the store?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/SingleCertLoader.cs [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+
+namespace Internal.Cryptography.Pal
+{
+    internal class SingleCertLoader : ILoaderPal
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/SingleCertLoader.cs]: sealed?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/SingleCertLoader.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        private ICertificatePal _cert;
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/SingleCertLoader.cs]: readonly?

**Comment by stephentoub** [comment]
A few nits, otherwise LGTM.

File: src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs (hunk 9 above)
Comment (bartonjs) [inline src/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]: Github artifact.  I took the one I had, added the where clause, and changed Linq to use it.  Github thinks I moved that one and changed the words, despite them being 3 commits apart.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs (hunk 10 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs]: Nah, it gets nulled out in the method currently named LoadTo

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs (hunk 11 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs]: Is there actually a benefit to sealing internal types?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs (hunk 15 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: We skip the first one because it was provided by the user; they still (might) have references to the objects.  (The comments and asserts are about the "store 0 is special"ness)

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs (hunk 12 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs]: > Maybe "MoveTo" since this seems to have move semantics rather than copy semantics?

Ooh, I like that much better, thanks.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs (hunk 11 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs]: Sealed all the new types; then made a new commit to seal out the old types that could be.

The question really is "when there's no InternalsVisibleTo, shouldn't the compiler/runtime already apply any benefits of sealing to non-extended internal types?"  But maybe that's asking for too much magic in the compiler.

**Comment by bartonjs** [comment]
The OSX Release run ran fine, and then:

> 11:05:41 [xUnit] [ERROR] - The plugin hasn't been performed correctly: Problem on deletion
11:05:41 Build step 'Publish xUnit test result report' changed build result to FAILURE

**Comment by bartonjs** [comment]
Test Innerloop OSX Release Build and Test please (#9066)

**Comment by bartonjs** [comment]
Test Innerloop OSX Release Build and Test please (#9066 again)

**Comment by bartonjs** [comment]
Okay, that's three times that the OSX Release build has passed, but failed to report it.  I'm just going to merge.

Thanks for the review, guys; I'm much happier with where this ended up.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs (hunk 11 above)
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertCollectionLoader.cs]: From my perspective there are three main reasons to mark a type as sealed:
1. Prevent unexpected usage from others deriving from the type.
2. Enable better performance by allowing for better devirtualization by the backend compiler.
3. Enable better code understandability; you look at a non-null variable of type X, and if it's sealed, you know that the instance referenced by that variable is definitively of that type and not something derived from it that may be overriding functionality, re-implementing an interface, adding more fields to make the instance more heavyweight, etc.

For an internal type, (1) isn't so much of an issue, and theoretically a compiler could do (2), but regardless (3) still applies.
